---
title: 'RNAmodR: RiboMethSeq'
author: "Felix G.M. Ernst"
date: "`r Sys.Date()`"
abstract: >
  detection of 2'-O methylations by RiboMethSeq
output:
  BiocStyle::html_document:
    toc: true
    toc_float: true
    df_print: paged
vignette: >
  %\VignetteIndexEntry{RNAmodR.RiboMethSeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: references.bib
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown(css.files = c('custom.css'))
```

# Introduction




```{r, echo = FALSE}
suppressPackageStartupMessages({
  library(GenomicFeatures)
  library(RNAmodR.RiboMethSeq)
})
```
```{r, eval = FALSE}
library(GenomicFeatures)
library(RNAmodR.RiboMethSeq)
```

# Example workflow

```{r}
gff <- system.file("extdata","example_RMS.gff3" ,package = "RNAmodR.Data")
fasta <- system.file("extdata","example_RMS.fasta" ,package = "RNAmodR.Data")
files <- list(c(treated = system.file("extdata","example_RMS.bam" ,package = "RNAmodR.Data")),
              c(treated = system.file("extdata","example_RMS.bam" ,package = "RNAmodR.Data")),
              c(treated = system.file("extdata","example_RMS.bam" ,package = "RNAmodR.Data")))
```

Analysis of data

```{r}
msrms <- ModSetRiboMethSeq(files,fasta,gff)
```

# Visualizing the results

To compare samples, we need to know, which positions should be part of the
comparison. This can either be done by aggregating the detect over all samples
and use the union or intersect or by using publish data. We want to assemble
a GRanges object from the latter. 

We will load a table and assemble the GRanges object.

```{r}
table <- read.csv2(system.file("extdata","snoRNAdb.csv",
                               package = "RNAmodR.Data"),
                   stringsAsFactors = FALSE)
table[1:2,]
# keep only the current coordinates
table <- table[,1:7]
gr <- GRanges(seqnames = table$hgnc_symbol,
              ranges = IRanges(start = table$position, width = 1),strand = "+",
              type = "RNAMOD",
              mod = table$modification,
              Parent = table$hgnc_symbol,
              Activity = CharacterList(strsplit(table$guide,",")))
# convert to current gene name
gr <- gr[vapply(gr$Activity != "unknown",all,logical(1)),]
head(coord)
coord <- split(gr,gr$Parent)
```

Heatmap

```{r}

```

sequence specific result

```{r}

```

# Session info

```{r, sessioninfo}
sessioninfo::session_info()
```

# References